Simplify before testing convergence

AnimeshSinha1309 opened the pull request:
Added simplification of function, before testing for convergence and added tests.

#### References to other Issues or PRs
Fixes #16699

#### Brief description of what is fixed or changed
Fixing issues with Sum of the factorial series, just added simplification of the expression before the tests for convergence start.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* concrete
  * simplify sums before testing for convergence
<!-- END RELEASE NOTES -->

AnimeshSinha1309 committed:
Simplify before testing convergence

Added simplification of function, before testing for convergence and added tests. Fixed #16699.

smichr commented:
Is there a way to just return None if you know that simplification? Or if this is a high-level routine then maybe using simplify makes sense.

maurogaravello commented:
I do not think this resolves the issue. This patch works in the specific example and probably is better than nothing.
If you consider 
```
expr = 1 - cos(1/n)
s = Sum(expr, (n, 1, oo))
```
the result given by SymPy is `False`, even with the proposed patch, but the correct result is `True`.
This is due, as in the original example, to a non sharp result of order.

A possible alternative is to evaluate `limit(expr * n, n, oo)`. In the case the result is a positive number (not 0), then the convergence is `False`. This is in substitution of the lines (445-446)
```
if p1_series_test[p] >= -1:
                return S.false 
```

AnimeshSinha1309 commented:
@smichr I did not fully understand what you said. Do you mean that if the call to simplify fails? When exactly should I return None.

My concern with calling simplify here was that the code might get slowed down, I think that this is becoming one of the slowest tests in the test suite. I am will try a smarter workaround if that is a problem, possibly changing the `__new__()` function in the Order class.

AnimeshSinha1309 commented:
Also the following calls to the order function still don't give answer that make sense:
```python
>>> print(O(factorial(n), (n, oo)))
O(factorial(n), (n, oo))
>>> print(O(factorial(n)/factorial(n-2), (n, oo)))
O(1, (n, oo))
>>> print(O(factorial(n)/factorial(n+2), (n, oo)))
O(1, (n, oo))
```
I would want that the `factorial()` calls all mapped to a result `O(exp(n*log(n)), (n, oo))`, because the factorial is bounded by `(n/2)**n` and `(n/3)**n`.

AnimeshSinha1309 commented:
So again, this function call is the problem here, right?
```python
>>> n = Symbol('n')
>>> O(1-cos(n), (n, 0))
O(1)
```
The result should have been `O(n^2)`, but that would require the Order function to have some understanding of the Taylor series expansions, won't it?

Also calling `limit(expr * n)` might just work, but even then all the calls to order don't really make much sense. :cry:

maurogaravello commented:
> So again, this function call is the problem here, right?
> 
> ```python
> >>> n = Symbol('n')
> >>> O(1-cos(n), (n, 0))
> O(1)
> ```

Exactly. This is the main problem. 
`O(1)` here is correct, but not sharp!
`O(n**-2)` is also correct. This would produce the correct result.

smichr commented:
I see that I had an incom plete thought in my suggestion. What it was intended to suggest is that maybe obtaining a meaningless O expression of something like that would allow you to know that you should return none of else use a limit or simplification.

AnimeshSinha1309 committed:
Revert "Simplify before testing convergence"

This reverts commit c46395e73b9d2389080e2b5f85b08ec302805885.

AnimeshSinha1309 committed:
Merge branch 'master' into issue/convergence-16699

Update the master to the current state and restart on the issue.

oscarbenjamin commented:
What's the status of this? @AnimeshSinha1309 @smichr

AnimeshSinha1309 commented:
Sorry, I did not follow up on this. My Attempts did not work out, I would appriciate any help I can get. I am available to work again, trying now.

AnimeshSinha1309 commented:
Honestly I am out of ideas, simple fixes to the `is_convergent()` method, or calling Simplify, fail on examples like `1-cos(1/n)`, and fixing the Order constructor is hard and I always end up breaking a bunch of tests. I am wondering if it would be okay to just fix a few cases, and leave fixing the Order function for another issue.

AnimeshSinha1309 committed:
Fixed issue by limit-comparison with P-series

Changed the is_convergent() method to test limit comparison against 1/n and 1/n**1.0001, fixing the convergence test problem, though not the sharp order problem.

AnimeshSinha1309 committed:
Typo fix

The previous commit forgot to call the function is_convergent in the tests, missing out the brackets. Fixing that here.

AnimeshSinha1309 commented:
https://github.com/sympy/sympy/blob/12d8f5354db591ba6600c50f4188b7b973a859fc/sympy/concrete/products.py#L359-L360
Is this example not incorrect, since `Product(n/(n + 1), (n, 1, oo)) = (1/2) * (2/3) * ... * (n-1/n) = 1/n = 1/oo = 0`. So the answer should be True, it's convergent, should it not?

AnimeshSinha1309 committed:
Fixing for Tests

Some tests fixed, some code. Changed some tests that seem most certainly wrong, esp. an example.

jksuom commented:
An infinite product with limit 0 is considered divergent.

AnimeshSinha1309 committed:
Both sides of 0

Handling the negative constant, fixing what I thought was an erroneous test and changed.

AnimeshSinha1309 commented:
Thanks for that, I fixed my mistake, undid the example that I had changed. I am just confused about one more thing, I changed a few lines in Summation, and added 3 tests for it, why is my Codecov score still falling, should I just keep adding mode examples?

AnimeshSinha1309 committed:
Fix a test

Issue #14488 is no longer raising a NotImplementedError due to the limit being evaluated correctly.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -496,12 +513,6 @@ def is_convergent(self):
-        ### ------------- alternating series test ----------- ###
-        dict_val = sequence_term.match((-1)**(sym + p)*q)
-        if not dict_val[p].has(sym) and is_decreasing(dict_val[q], interval):
-            return S.true
```
It looks like this and everything below is untested. Are these tests (alternating series etc) not already implemented somewhere else?

AnimeshSinha1309 replied:
I don't fully understand, there are already tests for Alternating Series and for everything else that follows. I believe that the issue might be that my p-series limit comparison is just returning a solution, blocking access to some of the functions later in the code. But now the codecov score looks pretty decent, right? And if not, what is it that I fix, I have no clue about codecov?

czgdp1807 commented:
Closing as the associated issue is outdated.
